content/en/blog/2022/rust-instrumentation/index.md

```diff
@@ -0,0 +1,396 @@
+
+## OpenTelemetry and Distributed Tracing in Rust
+
+### Tracing in Rust
+
+In Rust we have a great crate simply called [tracing](https://docs.rs/tracing/latest/tracing/). It is our core framework for tracing a Rust program. The crates documentation outlines its core concepts and APIs which are composed of these terms - spans, events, and subscribers.
+
+To record spans and events we use macros, here are some examples:
+
+```Rust
+{
+    // constructs a span with TRACE level.
+    let span = span!(Level::TRACE, "my span");
+    let _enter_guard = span.enter();
+
+    // recording an INFO level event
+    event!(Level::INFO, "inside the block");
+
+    // the _enter_guard will be dropped here and the span will exit.
+}
+{
+    let some_variable = 100_000;
+    // the entered consumes the span. which means it now owns it instead of just borrowing it.
+    let _enter_guard = info_span!("INFO level span").enter();
+
+    // records an ERROR level event with some attributes
+    error!(it = "accepts multiple arguments", and_also = ("multiple types"), how_many="?", could_be_more_than=some_variable);
+}
+```
+
+Subscriber - For spans and events to be recorded and aggregated we need to initialize a Subscriber. In most cases, we will use an existing Subscriber implementation.
+
+The crate [tracing_subscriber](https://docs.rs/tracing-subscriber/0.3.15/tracing_subscriber/) provides us with general use case subscriber that can be extended and modified according to specific needs.
+
+We initialize the subscriber using Registry struct:
+
+```shell
+let subscriber = tracing_subscriber::Registry::default()
+```
+
+This struct implements a subscriber and exposes us to another important feature of this crate - The Layer. Using layers, we can configure our subscriber to apply specific behaviors when interacting with spans and events.
+
+For example, if we want to filter some of our data, export it, format it or edit it, we can create or use an existing layer and compose it with the registry subscriber.
+
+```shell
+let subscriber = tracing_subscriber::Registry::default().with(SomeLayer)
+```
+
+We will see more examples on how to use these crates to instrument a Rust program in the practical section.
+
+### OpenTelemetry in Rust
+
+The crate that provides support for OpenTelemetry is called - how convenient - [OpenTelemetry](https://docs.rs/opentelemetry/latest/opentelemetry/).
+
+
+Let’s go over some of its key APIs:
+
+- Tracer. Inside the crate, there is a module called traces. It introduces us to a trait called Tracer. This is what tracks and connects our spans to create traces. To use Tracer, we create a subscriber layer. This is how the subscriber knows where to send the spans and how to generate traces.
+
+- Global. This module provides an API so that no matter in what section of the code we currently are, we can access the subscriber, tracer, and propagation context.
```

Comment:
```suggestion
- **Global**. This module provides an API so that no matter in what section of the code we currently are, we can access the subscriber, tracer, and propagation context.
```